tests/pyfunc/test_spark.py

```diff
@@ -309,6 +309,199 @@ def predict(self, context, model_input):
+            return {
+                "r1": [1] * input_len,
+                "r2": [1.5] * input_len,
+                "r3": [True] * input_len,
+                "r4": ["abc"] * input_len,
+            }
+
+    with mlflow.start_run() as run:
+        mlflow.pyfunc.log_model("model", python_model=TestModel(), signature=ModelSignature(
+            inputs=Schema([ColSpec("long")]),
+            outputs=Schema([
+                ColSpec('integer', 'r1'),
+                ColSpec('float', 'r2'),
+                ColSpec('boolean', 'r3'),
+                ColSpec('string', 'r4'),
+            ])
+        ))
+
+        udf = mlflow.pyfunc.spark_udf(
+            spark,
+            f"runs:/{run.info.run_id}/model",
+        )
+
+        data1 = spark.range(2).repartition(1)
+        result_spark_df = (
+            data1.withColumn("res", udf("id"))
+            .select("res.r1", "res.r2", "res.r3", "res.r4")
+        )
+        assert (
+            result_spark_df.schema.simpleString() == "struct<r1:int,r2:float,r3:boolean,r4:string>"
+        )
+
+        result = result_spark_df.toPandas()
+        assert result["r1"].tolist() == [1] * 2
+        np.testing.assert_almost_equal(result["r2"].tolist(), [1.5] * 2)
+        assert result["r3"].tolist() == [True] * 2
+        assert result["r4"].tolist() == ["abc"] * 2
+
+
+def test_spark_udf_tensorspec_struct_return_type_inference(spark):
+    class TestModel(PythonModel):
+        def predict(self, context, model_input):
+            input_len = len(model_input)
+            return {
+                "r1": [[1, 2]] * input_len,
+                "r2": [np.array([1.5, 2.5])] * input_len,
+                "r3": np.vstack([np.array([1.5, 2.5])] * input_len),
+                "r4": [np.array([[1., 2., 3.], [4., 5., 6.]])] * input_len,
+                "r5": np.array([[[1., 2., 3.], [4., 5., 6.]]] * input_len),
+            }
+
+    with mlflow.start_run() as run:
+        mlflow.pyfunc.log_model("model", python_model=TestModel(), signature=ModelSignature(
+            inputs=Schema([ColSpec("long")]),
+            outputs=Schema([
+                TensorSpec(np.dtype(np.int64), (2,), 'r1'),
+                TensorSpec(np.dtype(np.float64), (2,), 'r2'),
+                TensorSpec(np.dtype(np.float64), (2,), 'r3'),
+                TensorSpec(np.dtype(np.float64), (2, 3,), 'r4'),
+                TensorSpec(np.dtype(np.float64), (2, 3,), 'r5'),
+            ])
+        ))
+
+        udf = mlflow.pyfunc.spark_udf(
+            spark,
+            f"runs:/{run.info.run_id}/model",
+        )
+
+        data1 = spark.range(2).repartition(1)
+        result_spark_df = (
+            data1.withColumn("res", udf("id"))
+            .select("res.r1", "res.r2", "res.r3", "res.r4", "res.r5")
+        )
+        assert (
+            result_spark_df.schema.simpleString() ==
+            "struct<r1:array<bigint>,r2:array<double>,r3:array<double>,"
+            "r4:array<array<double>>,r5:array<array<double>>>"
```

Comment:
This isn't readable:

```suggestion
            "struct<r1:array<bigint>,"
            "r2:array<double>,"
            "r3:array<double>,"
            "r4:array<array<double>>,"
            "r5:array<array<double>>>"
```